Build fails with “'CGWindowListCreateImage' is unavailable: obsoleted in macOS 15.0”

eric-carlson opened the pull request:
#### 56fcab9726612823eaa685fd915bfb18673d6dcb
<pre>
Build fails with “&apos;CGWindowListCreateImage&apos; is unavailable: obsoleted in macOS 15.0”
<a href="https://bugs.webkit.org/show_bug.cgi?id=277564">https://bugs.webkit.org/show_bug.cgi?id=277564</a>
<a href="https://rdar.apple.com/133237307">rdar://133237307</a>

Reviewed by Jean-Yves Avenard

`CGWindowListCreateImage` is obsolete and unavailable in macOS 15, so write a method that
uses ScreenCaptureKit.framework to replace it.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/mac/ScreenCaptureKitSoftLink.h:
* Source/WebCore/PAL/pal/mac/ScreenCaptureKitSoftLink.mm:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
(WebCore::ScreenCaptureKitCaptureSource::captureWindowSnapshot):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _windowSnapshotInRect:]):
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(-[WKFullScreenWindowController enterFullScreen:]):
(takeWindowSnapshot):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::takeWindowSnapshot):
* Source/WebKitLegacy/mac/WebView/WebFullScreenController.mm:
(-[WebFullScreenController enterFullScreen:]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _windowSnapshotInRect:]):
* Source/WebKitLegacy/mac/WebView/WebViewPrivate.h:
* Tools/DumpRenderTree/mac/PixelDumpSupportMac.mm:
(takeWindowSnapshot):
(createBitmapContextFromWebView):
* Tools/MiniBrowser/mac/BrowserWindowController.h:
* Tools/MiniBrowser/mac/BrowserWindowController.m:
(-[BrowserWindowController windowSnapshotInRect:]):
(-[BrowserWindowController sharingService:transitionImageForShareItem:contentRect:]):
* Tools/MiniBrowser/mac/WK1BrowserWindowController.m:
(-[WK1BrowserWindowController windowSnapshotInRect:]):
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController windowSnapshotInRect:]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/mac/WebViewSnapshot.mm: Added.
(TestWebKitAPI::writeImage):
(TestWebKitAPI::TEST(WKWebView, WebViewSnapshot)):
* Tools/TestWebKitAPI/Tests/mac/checkered-background.html: Added.

Canonical link: <a href="https://commits.webkit.org/282351@main">https://commits.webkit.org/282351@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7b50ff7fb2077dd6b69dbe900f53f7d4a4a17308

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62863 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42219 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15459 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66883 "") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13467 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64983 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49906 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13751 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/66883 "") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9299 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65932 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39247 "Found 1 new test failure: compositing/geometry/fixed-position-composited-page-scale-smaller-than-viewport.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54455 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/66883 "") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35940 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11788 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12343 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/55973 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57487 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12118 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68578 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/62106 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6809 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11752 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58010 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6841 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54515 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58202 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13952 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5691 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/83869 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38039 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14756 "Found 1 new JSC stress test failure: wasm.yaml/wasm/fuzz/memory.js.wasm-no-cjit (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39119 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40230 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38861 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
```diff
@@ -651,6 +653,88 @@ - (void)outputVideoEffectDidStopForStream:(SCStream *)stream
+
+    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+
+    auto completionHandler = makeBlockPtr([&] (SCShareableContent* shareableContent, NSError* error) mutable {
```
I don't think we need to capture the environment by reference here.  

If needed for the logs, can we only capture what's actually needed here for clarity

jyavenard commented on Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.h:
```diff
@@ -75,6 +75,14 @@ class ScreenCaptureKitCaptureSource final
     static std::optional<CaptureDevice> screenCaptureDeviceWithPersistentID(const String&);
     static std::optional<CaptureDevice> windowCaptureDeviceWithPersistentID(const String&);
 
+    enum SnapshotOptions {
```
shouldn't this be an enum class ?
They are used with fully qualified name later.

jyavenard commented on Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
```diff
@@ -651,6 +653,88 @@ - (void)outputVideoEffectDidStopForStream:(SCStream *)stream
+            return;
+        }
+
+        SCWindow* targetWindow = nullptr;
```
`RetainPtr<SCWindow>` ?

eric-carlson replied:
`shareableContent.windows` is a an NSDictionary, which already has a retain on all of the windows and we won't use it outside of the lambda.

eric-carlson committed:
Build fails with “'CGWindowListCreateImage' is unavailable: obsoleted in macOS 15.0”
https://bugs.webkit.org/show_bug.cgi?id=277564
rdar://133237307

Reviewed by Jean-Yves Avenard

`CGWindowListCreateImage` is obsolete and unavailable in macOS 15, so write a method that
uses ScreenCaptureKit.framework to replace it.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/mac/ScreenCaptureKitSoftLink.h:
* Source/WebCore/PAL/pal/mac/ScreenCaptureKitSoftLink.mm:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
(WebCore::ScreenCaptureKitCaptureSource::captureWindowSnapshot):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _windowSnapshotInRect:]):
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(-[WKFullScreenWindowController enterFullScreen:]):
(takeWindowSnapshot):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::takeWindowSnapshot):
* Source/WebKitLegacy/mac/WebView/WebFullScreenController.mm:
(-[WebFullScreenController enterFullScreen:]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _windowSnapshotInRect:]):
* Source/WebKitLegacy/mac/WebView/WebViewPrivate.h:
* Tools/DumpRenderTree/mac/PixelDumpSupportMac.mm:
(takeWindowSnapshot):
(createBitmapContextFromWebView):
* Tools/MiniBrowser/mac/BrowserWindowController.h:
* Tools/MiniBrowser/mac/BrowserWindowController.m:
(-[BrowserWindowController windowSnapshotInRect:]):
(-[BrowserWindowController sharingService:transitionImageForShareItem:contentRect:]):
* Tools/MiniBrowser/mac/WK1BrowserWindowController.m:
(-[WK1BrowserWindowController windowSnapshotInRect:]):
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController windowSnapshotInRect:]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/mac/WebViewSnapshot.mm: Added.
(TestWebKitAPI::writeImage):
(TestWebKitAPI::TEST(WKWebView, WebViewSnapshot)):
* Tools/TestWebKitAPI/Tests/mac/checkered-background.html: Added.

Canonical link: https://commits.webkit.org/282351@main
